### PR TITLE
New storage layer: gcs(google cloud storage)

### PR DESCRIPTION
**Pull request by yinhm**

### CODE REVIEW

File: imagestore/gcsstore.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+type GCSImageStore struct {
+	ctx            context.Context
+	bucketName     string
+	storeRoot      string
+	namePathMapper *NamePathMapper
+}
+
+func NewGCSImageStore(ctx context.Context, bucket string, root string, mapper *NamePathMapper) *GCSImageStore {
+	log.Println("new gcs image store.")
```
Comment (briankassouf) [inline imagestore/gcsstore.go]: Could you remove this log line?

File: imagestore/gcsstore.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	_, err := storage.StatObject(this.ctx, this.bucketName, this.toPath(obj))
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (this *GCSImageStore) Save(src string, obj *StoreObject) (*StoreObject, error) {
+	log.Println("ready to save image.")
```
Comment (briankassouf) [inline imagestore/gcsstore.go]: Please remove log line

File: imagestore/gcsstore.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	wc.ContentType = obj.MimeType
+	if _, err := wc.Write(data); err != nil {
+		log.Printf("error on write data: %s", err)
+		return nil, err
+	}
+	if err := wc.Close(); err != nil {
+		log.Printf("error on close writer: %s", err)
+		return nil, err
+	}
+	log.Println("updated object:", wc.Object())
```
Comment (briankassouf) [inline imagestore/gcsstore.go]: Please remove log line

File: imagestore/gcsstore.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	obj.Url = "https://storage.googleapis.com/" + this.bucketName + "/" + this.toPath(obj)
+	log.Println("object url: ", obj.Url)
```
Comment (briankassouf) [inline imagestore/gcsstore.go]: Please remove log line

File: imagestore/gcsstore.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	return &GCSImageStore{
+		ctx:            ctx,
+		bucketName:     bucket,
+		storeRoot:      root,
+		namePathMapper: mapper,
+	}
+}
+
+func (this *GCSImageStore) Exists(obj *StoreObject) (bool, error) {
+	log.Println("checking exists: ", this.toPath(obj))
```
Comment (briankassouf) [inline imagestore/gcsstore.go]: Please remove log line

File: config/default.conf.json [hunk 6]
```diff
@@ -13,6 +13,15 @@
...
         },
         {
+            "Type" : "gcs",
+            "BucketName" : "",
+            "StoreRoot" : "",
+            "AppID" : "",
+            "KeyFile" : "appid.json",
+            "NamePathRegex" : "",
+            "NamePathMap" : "${ImageSize}/${ImageName}"
+        }
```
Comment (briankassouf) [inline config/default.conf.json]: You need a comma after this bracket '},'

**Comment by briankassouf** [comment]
@yinhm Thanks for submitting this Pull Request! having google cloud storage support will be very helpful for users!

**Comment by lospro7** [comment]
Other than comments :+1: @yinhm this is awesome. thanks for contributing.

**Comment by yinhm** [comment]
@briankassouf Thanks for the review, pull updated.

**Comment by yinhm** [comment]
@lospro7 Ofc, I have a naive testcase in my project: https://gist.github.com/yinhm/8fffe49f0dc637d8fef3

Somebody may found it useful if you are interested in an mandible go client example, it only do "/url" post for my current requirement.

**Pull request merged**